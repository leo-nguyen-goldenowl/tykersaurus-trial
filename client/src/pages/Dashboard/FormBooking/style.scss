@import "src/assets/scss/_variables.scss";
@import "src/assets/scss/_mixins.scss";
@import "src/assets/scss/_extends.scss";

.form-booking {
  position: relative;
  max-width: 1170px;
  padding: 40px;
  border-radius: 20px;
  background-color: $white;
  box-shadow: 0 2px 10px rgba(0, 0, 0, 0.2);
  &__top {
    margin-bottom: 30px;
    @include responsive(tablet) {
      @include flex(null,center);
    }
    @include responsive(small-desktop) {
      @include flex();
    }

    &__item {
      margin-right: 40px;
      @include multiple-responsive(small-tablet, mobile, small-mobile) {
        margin-top: 15px;
        margin-right: 0;
        &:first-child {
          margin-top: 0;
        }
      }
    }
  }
  &__bottom {
    margin-bottom: 30px;
    @include multiple-responsive(
      small-desktop,
      tablet,
      small-tablet,
      mobile,
      small-mobile
    ) {
      @include flex(space-between, center);
      flex-wrap: wrap;
    }
    @include responsive(medium-desktop) {
      @include flex(space-between);
    }

    &__item {
      @include multiple-responsive(tablet, small-tablet, mobile, small-mobile) {
        width: 100%;
        margin-bottom: 30px;
        &:last-child {
          margin-bottom: 0;
        }
      }

      @include responsive(small-desktop) {
        width: calc(100% / 2 - 15px);
        margin-bottom: 30px;
        &:last-child {
          margin-bottom: 0;
        }
      }
      @include responsive(medium-desktop) {
        width: calc(100% / 3 - 15px);
      }
    }
  }
  &__description {
    @include text(SFProDisplay-Regular, 14px, $midnight-moss, 1.43);
    margin-bottom: 20px;
    span {
      font-family: SFProDisplay-Bold;
    }
  }
  &__btn-submit {
    position: absolute;
    width: 370px;
    bottom: 0;
    left: 50%;
    transform: translate(-50%, 50%);
    @include multiple-responsive(small-mobile, mobile, small-tablet) {
      width: 200px;
    }
  }
}
