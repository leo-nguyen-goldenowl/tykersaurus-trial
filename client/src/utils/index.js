export const getValueFromStorage = (name) => {
  return localStorage.getItem(name)
}
