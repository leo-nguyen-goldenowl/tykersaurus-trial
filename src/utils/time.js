const moment = require('moment')

const convertTeeTimeToMinute = (teeTime) => {
  const listPartOfTeeTime = teeTime.split(':')
  const hourTeeTime = listPartOfTeeTime[0]
  const minuteTeeTime = listPartOfTeeTime[1]

  return Number(hourTeeTime * 60 + minuteTeeTime)
}

const getMaxDate = () => {
  const todayMoment = moment()
  return moment(todayMoment).add(
    moment(moment()).diff(
      moment(
        moment().add(-1, 'days').format('MM/DD/YYYY 23:00'),
        'MM-DD-YYYY hh:mm'
      ),
      'seconds'
    ) > 0
      ? 7
      : 6,
    'days'
  )
}

const accessBooking = (date) => {
  const maxDate = getMaxDate()

  return (
    moment(moment(maxDate).format('MM/DD/YYYY 00:00')).diff(
      moment(date),
      'seconds'
    ) >= 0
  )
}

module.exports = { convertTeeTimeToMinute, accessBooking, getMaxDate }
