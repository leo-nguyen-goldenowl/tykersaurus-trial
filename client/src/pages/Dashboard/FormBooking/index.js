import React from 'react'
import moment from 'moment'
import { useMergeState } from 'hooks'

import ButtonDefault from 'components/Button/Default'
import SelectMini from 'components/Select/Mini'
import SelectLarge from 'components/Select/Large'
import DatePicker from 'components/Picker/Date'
import TimePicker from 'components/Picker/Time'

import './style.scss'

const FormBooking = () => {
  const listSelectTop = [
    {
      label   : 'Session',
      name    : 'session',
      listItem: [
        { name: 'Morning', value: '07:00,11:59' },
        { name: 'Afternoon', value: '12:00,18:00' }
      ]
    },
    {
      label   : 'Hole',
      name    : 'hole',
      listItem: [
        { name: '9 Holes', value: 9 },
        { name: '18 Holes', value: 18 }
      ]
    },
    {
      label   : 'Quantity of Players',
      name    : 'player',
      listItem: [
        { name: '2 Players', value: 2 },
        { name: '3 Players', value: 3 },
        { name: '4 Players', value: 4 }
      ]
    }
  ]

  const [dataBooking, setDataBooking] = useMergeState({
    session     : '07:00,11:59',
    hole        : 9,
    player      : 2,
    course      : 'Classic Course',
    date        : moment().format('ddd, MMM DD YYYY'),
    teeTimeRange: {
      from: moment(),
      to  : moment()
    }
  })

  const handleChangeDate = (value) => {
    setDataBooking({ ['date']: moment(value).format('ddd, MMM DD YYYY') })
  }

  const handleChangeSelect = (e) => {
    const { name, value } = e.target
    setDataBooking({ [name]: value })
  }

  const handleChangeTime = (value, type) => {
    setDataBooking({
      teeTimeRange: {
        ...dataBooking.teeTimeRange,
        [type]: value
      }
    })
  }

  const handleValidDataBooking = () => {
    const { from: fromTeeTime, to: toTeetime } = dataBooking.teeTimeRange

    return (
      moment(moment(toTeetime)).diff(moment(moment(fromTeeTime)), 'minutes') >=
      0
    )
  }

  const handleSubmit = () => {
    if (handleValidDataBooking()) {
      console.log('Ready to book')
    } else {
      console.log('Invalid data')
    }
  }

  return (
    <div className='form-booking'>
      <div className='form-booking__top'>
        {listSelectTop.map((select, idx) => (
          <div className='form-booking__top__item' key={idx}>
            <SelectMini
              label={select.label}
              name={select.name}
              value={dataBooking[select.name]}
              listItem={select.listItem}
              onChange={handleChangeSelect}
            />
          </div>
        ))}
      </div>
      <div className='form-booking__bottom'>
        <div className='form-booking__bottom__item'>
          <SelectLarge
            label='Course'
            name='course'
            value={dataBooking.course}
            onChange={handleChangeSelect}
            listItem={[
              { name: 'Classic Course', value: 'Classic Course' },
              {
                name : 'Masters Course',
                value: 'Masters Course'
              }
            ]}
          />
        </div>
        <div className='form-booking__bottom__item'>
          <DatePicker
            label='Calendar'
            name='calendar'
            value={dataBooking.date}
            type='calendar'
            onChange={handleChangeDate}
          />
        </div>
        <div className='form-booking__bottom__item'>
          <TimePicker
            value={dataBooking.teeTimeRange}
            label='Time'
            onChange={handleChangeTime}
          />
        </div>
      </div>
      <p className='form-booking__description'>
        <span>Description:</span> Lorem ipsum dolor sit amet, modus inimicus eum
        no <span>7AM</span> at nam modus <span>1 week</span> consul. Mucius
        oportere vel id. Ut solum malorum conclusionemque qui. Mea te graeci
        eruditi honestatis. Ne iisque placerat est, mel in dicunt invenire. Te
        his delenit eloquentiam.
      </p>
      <div className='form-booking__btn-submit'>
        <ButtonDefault type='button' onClick={handleSubmit}>
          Book Now
        </ButtonDefault>
      </div>
    </div>
  )
}

export default FormBooking
