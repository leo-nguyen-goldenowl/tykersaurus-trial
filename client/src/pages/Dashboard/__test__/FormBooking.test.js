import { testSnapshots } from "utils/test/test.utils";
import FormBooking from "../FormBooking";

jest
  .mock("components/Button/Default", () => "ButtonDefault")
  .mock("components/Select/Mini", () => "SelectMini")
  .mock("components/Select/Large", () => "SelectLarge")
  .mock("components/Picker/Date", () => "DatePicker")
  .mock("components/Picker/Time", () => "TimePicker")
  .mock("react-redux", () => ({
    useDispatch: () => (args) => args,
  }))
  // .mock("moment", () => "moment")
  .mock("hooks", () => ({ useMergeState: () => "useMergeState" }))
  .mock("redux/services/ticket", () => ({
    bookingCourse: () => (args) => args,
  }));

describe("Test FormBooking component", () => {
  testSnapshots(FormBooking, [
    {
      props: {},
      description: "FormBooking component should match snapshots",
    },
  ]);
});
