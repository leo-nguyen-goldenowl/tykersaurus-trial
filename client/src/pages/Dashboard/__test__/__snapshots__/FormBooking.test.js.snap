// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test FormBooking component snapshots FormBooking component should match snapshots 1`] = `
<div
  className="form-booking"
>
  <div
    className="form-booking__top"
  >
    <div
      className="form-booking__top__item"
    >
      <SelectMini
        label="Session"
        listItem={
          Array [
            Object {
              "name": "Morning",
              "value": "07:00,11:59",
            },
            Object {
              "name": "Afternoon",
              "value": "12:00,18:00",
            },
          ]
        }
        name="session"
        onChange={[Function]}
      />
    </div>
    <div
      className="form-booking__top__item"
    >
      <SelectMini
        label="Hole"
        listItem={
          Array [
            Object {
              "name": "9 Holes",
              "value": 9,
            },
            Object {
              "name": "18 Holes",
              "value": 18,
            },
          ]
        }
        name="hole"
        onChange={[Function]}
      />
    </div>
    <div
      className="form-booking__top__item"
    >
      <SelectMini
        label="Quantity of Players"
        listItem={
          Array [
            Object {
              "name": "2 Players",
              "value": 2,
            },
            Object {
              "name": "3 Players",
              "value": 3,
            },
            Object {
              "name": "4 Players",
              "value": 4,
            },
          ]
        }
        name="player"
        onChange={[Function]}
      />
    </div>
  </div>
  <div
    className="form-booking__bottom"
  >
    <div
      className="form-booking__bottom__item"
    >
      <SelectLarge
        label="Course"
        listItem={
          Array [
            Object {
              "name": "Classic Course",
              "value": "Classic Course",
            },
            Object {
              "name": "Masters Course",
              "value": "Masters Course",
            },
          ]
        }
        name="course"
        onChange={[Function]}
      />
    </div>
    <div
      className="form-booking__bottom__item"
    >
      <DatePicker
        label="Calendar"
        name="calendar"
        onChange={[Function]}
        type="calendar"
      />
    </div>
    <div
      className="form-booking__bottom__item"
    >
      <TimePicker
        label="Time"
        onChange={[Function]}
      />
    </div>
  </div>
  <p
    className="form-booking__description"
  >
    <span>
      Description:
    </span>
     Lorem ipsum dolor sit amet, modus inimicus eum no 
    <span>
      7AM
    </span>
     at nam modus 
    <span>
      1 week
    </span>
     consul. Mucius oportere vel id. Ut solum malorum conclusionemque qui. Mea te graeci eruditi honestatis. Ne iisque placerat est, mel in dicunt invenire. Te his delenit eloquentiam.
  </p>
  <div
    className="form-booking__btn-submit"
  >
    <ButtonDefault
      onClick={[Function]}
      type="button"
    >
      Book Now
    </ButtonDefault>
  </div>
</div>
`;
